[android] Expose setNumThreads to android api

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31033 [android] Expose setNumThreads to android api**

Intention:
There are requests from users to control number of threads from android side:
https://discuss.pytorch.org/t/android-pytorch-forward-method-running-in-a-separate-thread-slow-down-ui-thread/63516/2
https://discuss.pytorch.org/t/threading-of-model-pytorch-android/62490/2

At the moment `setNumThreads` is placed in `org.pytorch.Module`, but this method changes global threadPool size, in future we will move it to some separate class to repeat python binding structure, which has torch.set_num_threads()

Differential Revision: [D18923167](https://our.internmc.facebook.com/intern/diff/D18923167)

IvanKobzarev committed:
[android] Expose setNumThreads to android api

IvanKobzarev committed:
Update on "[android] Expose setNumThreads to android api"




[ghstack-poisoned]

dreiss commented:
Looks good to me.  Can you update the commit message with more information about why we're adding this, and add javadoc to explain how it's supposed to be used?

IvanKobzarev committed:
Update on "[android] Expose setNumThreads to android api"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Expose setNumThreads to android api"


Intention:
There are requests from users to control number of threads from android side:
https://discuss.pytorch.org/t/android-pytorch-forward-method-running-in-a-separate-thread-slow-down-ui-thread/63516/2
https://discuss.pytorch.org/t/threading-of-model-pytorch-android/62490/2

At the moment `setNumThreads` is placed in `org.pytorch.Module`, but this method changes global threadPool size, in future we will move it to some separate class to repeat python binding structure, which has torch.set_num_threads()


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Expose setNumThreads to android api"


Intention:
There are requests from users to control number of threads from android side:
https://discuss.pytorch.org/t/android-pytorch-forward-method-running-in-a-separate-thread-slow-down-ui-thread/63516/2
https://discuss.pytorch.org/t/threading-of-model-pytorch-android/62490/2

At the moment `setNumThreads` is placed in `org.pytorch.Module`, but this method changes global threadPool size, in future we will move it to some separate class to repeat python binding structure, which has torch.set_num_threads()


[ghstack-poisoned]

dreiss commented on caffe2/utils/threadpool/ThreadPool.h:
```diff
@@ -49,9 +51,9 @@ class CAFFE2_API /*alignas(kCacheLineSize)*/ ThreadPool {
 
  private:
-  mutable std::mutex executionMutex_;
+  std::mutex executionMutex_;
```
Maybe revert this change or move it to another diff?  It seems unrelated.

IvanKobzarev committed:
Update on "[android] Expose setNumThreads to android api"


Intention:
There are requests from users to control number of threads from android side:
https://discuss.pytorch.org/t/android-pytorch-forward-method-running-in-a-separate-thread-slow-down-ui-thread/63516/2
https://discuss.pytorch.org/t/threading-of-model-pytorch-android/62490/2

At the moment `setNumThreads` is placed in `org.pytorch.Module`, but this method changes global threadPool size, in future we will move it to some separate class to repeat python binding structure, which has torch.set_num_threads()


[ghstack-poisoned]
